src/main/java/software/amazon/encryption/s3/internal/CipherSubscriber.java

```diff
@@ -91,21 +125,63 @@ public void onError(Throwable t) {
 
     @Override
     public void onComplete() {
+        // In rare cases, e.g. when the last part of a low-level MPU has 0 length,
+        // onComplete will be called before onNext is called once.
+        if (contentRead.get() < contentLength) {
```

Comment:
Should this calculation include the tag length?
It feels weird that this formula is different than the other ones
```suggestion
        if (contentRead.get() + tagLength < contentLength) {
```